### PR TITLE
sending and receiving data from esp32 to cc1312 radio module

### PR DESCRIPTION
**Pull request by gustavosinbandera1**

### CODE REVIEW

File: components/HTTP/src/RESTServer.cpp [hunk 1]
```diff
@@ -431,21 +432,30 @@ esp_err_t websocketHandler(httpd_req_t* req)
...
+    ws_instance.onReceive(ws_pkt, req);
     return ret;
 }
 
+/**
+ * @brief this function aims to setting up the api rest end points,
+ * start the server amd initialize Radio module serial communication
+ * 
+ */
+
```
Comment (jeandudey) [inline components/HTTP/src/RESTServer.cpp:444]: Move to `RESTServer.h` file please.

File: components/HTTP/src/RESTServer.h [hunk 2]
```diff
@@ -15,16 +15,18 @@
 #include <cstdint>
 #include <cstring>
 
+
+
```
Comment (jeandudey) [inline components/HTTP/src/RESTServer.h:19]:
```suggestion
```

Unrelated changes.

File: components/HTTP/src/RESTServer.h [hunk 3]
```diff
@@ -15,16 +15,18 @@
...
 namespace rest_server {
 
 /**
  * @brief Start the request server in a given port
  * 
  * @param port 
  */
+
```
Comment (jeandudey) [inline components/HTTP/src/RESTServer.h:27]: Don't space up the doc comment between the function.

File: components/HTTP/src/Websocket.cpp [hunk 4]
```diff
@@ -22,6 +22,37 @@ Websocket::Websocket()
     g_check_connections.start(NULL);
 }
 
+ /**
+     * @brief received encrypted message by radio
+     * @param data_received encrypted message
+*/
```
Comment (jeandudey) [inline components/HTTP/src/Websocket.cpp:28]:
```suggestion
/**
 * @brief received encrypted message by radio
 * @param data_received encrypted message
 */
```

File: components/HTTP/src/Websocket.cpp [hunk 5]
```diff
@@ -22,6 +22,37 @@ Websocket::Websocket()
...
+        }
+
+        httpd_ws_frame_t ws_pkt;
+        memset(&ws_pkt, 0, sizeof(httpd_ws_frame_t));
+        ws_pkt.payload = (std::uint8_t*)payload;
+        ws_pkt.type = HTTPD_WS_TYPE_TEXT;
+        ws_pkt.len = strlen(payload);
+        ws_pkt.final = 1;
+       
+    // send message to customers
```
Comment (jeandudey) [inline components/HTTP/src/Websocket.cpp:46]: Four lines indentation.

File: components/HTTP/src/Websocket.cpp [hunk 6]
```diff
@@ -22,6 +22,37 @@ Websocket::Websocket()
...
 
+ /**
+     * @brief received encrypted message by radio
+     * @param data_received encrypted message
+*/
+static void receiveFromUart(void* data_received, void* length)
+{
+        char payload[500];
+        esp_err_t err;
+        fwrite(data_received, (size_t)length, 1, stdout);
```
Comment (jeandudey) [inline components/HTTP/src/Websocket.cpp:33]: Please use ESP-IDF log functions, or hexdump log functions for this purpose so this can be deactivated when unused.

File: components/HTTP/src/Websocket.h [hunk 7]
```diff
@@ -67,7 +70,9 @@ class Websocket
 
     Websocket& operator=(Websocket const&) = delete; // Copy assign
     Websocket& operator=(Websocket&&) = delete;      // Move assign
+    void initRadioSerialLine(void);
 
+
```
Comment (jeandudey) [inline components/HTTP/src/Websocket.h:75]:
```suggestion
```

Trailing whitespaces

File: components/Radio/src/Radio.cpp [hunk 8]
```diff
@@ -11,58 +11,109 @@
...
 #include "Radio.h"
 
+#include "freertos/queue.h"
 #include <driver/uart.h>
 #include <esp_log.h>
+#include <string.h>
+
+
+
+#define FTDI_PORT UART_NUM_0
```
Comment (jeandudey) [inline components/Radio/src/Radio.cpp:21]: We can't use `UART_NUM_0` while testing cause it's used by the FTDI and for the stdout (serial console), use `UART_NUM_1` or `UART_NUM_2`. Can the `#define` be moved to `defaults.h`?

File: components/Radio/src/Radio.cpp [hunk 9]
```diff
@@ -11,58 +11,109 @@
...
+#define BUF_SIZE (1024)
+#define RD_BUF_SIZE (BUF_SIZE)
 
 namespace radio {
 
 #if RAD_ENABLED == true
 
 static const char* TAG = "UART";
+static QueueHandle_t uart0_queue;
```
Comment (jeandudey) [inline components/Radio/src/Radio.cpp:30]:
```suggestion
static QueueHandle_t g_uart0_queue;
```

File: components/Radio/src/Radio.cpp [hunk 10]
```diff
@@ -11,58 +11,109 @@
...
 
 Radio::Radio()
     : Task("UART", 4096, 5)
 {
 }
 
-void Radio::run(void* data)
+void Radio::init(callbackFunction fn)
```
Comment (jeandudey) [inline components/Radio/src/Radio.cpp:37]: Give `callbackFunction` a more meaningful name and follow naming guidelines, like `radio_rx_cb_t`.

File: components/Radio/src/Radio.cpp [hunk 11]
```diff
@@ -11,58 +11,109 @@
...
+void Radio::sendDataToRadio(void* data, size_t length)
+{
+    uart_write_bytes(FTDI_PORT, (const char*) data, length);
+}
+
+void Radio::run(void* data)
+{   
+    uart_event_t event;
+    size_t buffered_size;
+    uint8_t* dtmp = (uint8_t*) malloc(RD_BUF_SIZE);
```
Comment (jeandudey) [inline components/Radio/src/Radio.cpp:74]: This doesn't need to be allocated with `malloc`, it can live on the stack.

Comment (jeandudey) [reply components/Radio/src/Radio.cpp:74]:
```suggestion
    uint8_t dtmp[RD_ BUF_SIZE];
```

Reason: the size is constant, does not change.

File: components/Radio/src/Radio.h [hunk 12]
```diff
@@ -18,6 +18,8 @@
 namespace radio {
 
 #if RAD_ENABLED == true
+//this type of data represent the callback to send data to websocket server
+typedef void (*callbackFunction)(void*,void*);
```
Comment (jeandudey) [inline components/Radio/src/Radio.h:22]: See above comment about renaming this.

File: components/Radio/src/Radio.h [hunk 13]
```diff
@@ -31,6 +33,20 @@ class Radio : public Task
...
+    //in order to receive the data from serial port
+    /**
+     * @brief This function need to receive the callback to return the data from UART interface
+     * 
+     * @param fn callback function to receive data out of this class
+     * void myCallback(void* buffer, void* size)
+     * {
+     *  //some implementation
+     * }
+     */
```
Comment (jeandudey) [inline components/Radio/src/Radio.h:46]: Use correct doc comments.

File: components/Radio/src/Radio.h (hunk 12 above)
Comment (jeandudey) [reply components/Radio/src/Radio.h:22]: Use correct doxygen doc comment.

File: components/Radio/src/Radio.cpp (hunk 8 above)
Comment (gustavosinbandera1) [reply components/Radio/src/Radio.cpp:21]: yes, the port can be moved and the definition as well

File: components/HTTP/src/Websocket.cpp [hunk 14]
```diff
@@ -22,6 +22,37 @@ Websocket::Websocket()
...
+        err = util::decode((std::uint8_t*)data_received, payload);
+        if (err) {
+            ESP_LOGI("TEST", "error decoding the value");
```
Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:36]: Use ESP_LOGE for error

File: components/Radio/src/Radio.cpp [hunk 15]
```diff
@@ -11,58 +11,109 @@
...
+                    break;
+                //Event of HW FIFO overflow detected
+                case UART_FIFO_OVF:
+                    ESP_LOGI(TAG, "hw fifo overflow"); 
+                    uart_flush_input(FTDI_PORT);
+                    xQueueReset(uart0_queue);
+                    break;
+                //Event of UART ring buffer full
+                case UART_BUFFER_FULL:
+                    ESP_LOGI(TAG, "ring buffer full");
```
Comment (luisan00) [inline components/Radio/src/Radio.cpp:92]: This should be log warning or log error

File: components/Radio/src/Radio.cpp [hunk 16]
```diff
@@ -11,58 +11,109 @@
...
+                    uart_flush_input(FTDI_PORT);
+                    xQueueReset(uart0_queue);
+                    break;
+                //Event of UART RX break detected
+                case UART_BREAK:
+                    ESP_LOGI(TAG, "uart rx break");
+                    break;
+                //Event of UART parity check error
+                case UART_PARITY_ERR:
+                    ESP_LOGI(TAG, "uart parity error");
```
Comment (luisan00) [inline components/Radio/src/Radio.cpp:102]: Change info log to warning or error log

File: components/Radio/src/Radio.cpp [hunk 17]
```diff
@@ -11,58 +11,109 @@
...
+                    break;
+                //Event of UART frame error
+                case UART_FRAME_ERR:
+                    ESP_LOGI(TAG, "uart frame error");
```
Comment (luisan00) [inline components/Radio/src/Radio.cpp:106]: Should be ESP_LOGE

File: components/Radio/src/Radio.cpp [hunk 18]
```diff
@@ -11,58 +11,109 @@
...
+        if(xQueueReceive(uart0_queue, (void * )&event, (portTickType)portMAX_DELAY)) {
+            bzero(dtmp, RD_BUF_SIZE);
+            switch(event.type) {
+                case UART_DATA:
+                    uart_read_bytes(FTDI_PORT, dtmp, event.size, portMAX_DELAY);
+                    callback_handler((void*)dtmp,(void*)event.size);
+                    break;
+                //Event of HW FIFO overflow detected
+                case UART_FIFO_OVF:
+                    ESP_LOGI(TAG, "hw fifo overflow");
```
Comment (luisan00) [inline components/Radio/src/Radio.cpp:86]: Should be a warning log, not info

File: components/HTTP/src/RESTServer.cpp (hunk 1 above)
Comment (jeandudey) [reply components/HTTP/src/RESTServer.cpp:444]: Did you move this?

Comment (gustavosinbandera1) [reply components/HTTP/src/RESTServer.cpp]: yes
